pin-project-internal/src/project.rs

```diff
@@ -171,17 +172,31 @@ fn replace_item_impl(item: &mut ItemImpl, mutability: Mutability) {
         .for_each(|item| determine_lifetime_name(&mut lifetime_name, &item.sig.generics.params));
     let lifetime = Lifetime::new(&lifetime_name, Span::call_site());
 
-    insert_lifetime(&mut item.generics, lifetime.clone());
+    let generics: Generics = syn::parse_quote!(#arguments);
 
     match arguments {
         PathArguments::None => {
             *arguments = PathArguments::AngleBracketed(syn::parse_quote!(<#lifetime>));
         }
         PathArguments::AngleBracketed(args) => {
+            for arg in &args.args {
+                if let GenericArgument::Lifetime(life) = arg {
+                    if life.ident == "_" {
+                        return Err(error!(
+                            life,
+                            "#[project] attribute may not be used with '_ lifetimes"
+                        ));
+                    }
+                }
+            }
             args.args.insert(0, syn::parse_quote!(#lifetime));
         }
         PathArguments::Parenthesized(_) => unreachable!(),
     }
+
+    let pred = insert_lifetime_and_bound(&mut item.generics, lifetime, &generics, orig_ident);
```

Comment:
I noticed that as well when I was writing this - however, I couldn't figure out exactly why it was allowed. I thought that it was better to include the bound on the impl, just in case there's there's a corner case which does require it.

Reply:
Maybe: https://github.com/rust-lang/rfcs/blob/master/text/2093-infer-outlives.md#background-outlives-requirements-today

> This is due to Rust's support for implied bounds -- in particular, every function and impl assumes that the types of its inputs are well-formed. In this case, since foo can assume that &'a T is well-formed, it can also deduce that T: 'a must hold, and hence we do not require where-clauses asserting this fact.

Reply:
I see - I didn't realize that this applied to impls.